elki-outlier/src/main/java/elki/outlier/isolation/IsolationForest.java

```diff
@@ -0,0 +1,366 @@
+
+    return FastMath.pow(2, -avgPathLength / c(subsampleSize));
+
+    // return scores;
+  }
+
+  /**
+   * Inner class for building and using an isolation tree.
+   */
+  static class Tree {
+
+    /**
+     * The size of the node
+     */
+    private int size;
+
+    /**
+     * The split attribute
+     */
+    private int a;
+
+    /**
+     * The split point
+     */
+    private double q;
+
+    /**
+     * The successors
+     */
+    private List<Tree> successors;
+
+    /**
+     * Constructs a tree from data.
+     *
+     * @param data Subsample from data
+     * @param r Random generator
+     * @param height
+     * @param maxHeight
+     */
+    protected Tree(ModifiableDBIDs data, Random r, int height, int maxHeight, Relation<? extends NumberVector> relation) {
+      // Set size of node
+      size = data.size();
+
+      // Stop splitting if necessary
+      if((size <= 1) || (height == maxHeight)) {
+        return;
+      }
+
+      // Compute mins and maxs and eligible attributes
+      final int dim = RelationUtil.dimensionality(relation);
+      double[][] minmax;
+      {
+        double[] mins = new double[dim], maxs = new double[dim];
+        for(int i = 0; i < dim; i++) {
+          mins[i] = Double.MAX_VALUE;
+          maxs[i] = -Double.MAX_VALUE;
+        }
+        for(DBIDIter iditer = data.iter(); iditer.valid(); iditer.advance()) {
+          final NumberVector o = relation.get(iditer);
+          for(int d = 0; d < dim; d++) {
+            final double v = o.doubleValue(d);
+            mins[d] = Math.min(v, mins[d]);
+            maxs[d] = Math.max(v, maxs[d]);
+          }
+        }
+        minmax = new double[][] { mins, maxs };
+      }
+
+      ArrayList<Integer> al = new ArrayList<Integer>();
+      for(int j = 0; j < dim; j++) {
+        if(minmax[0][j] < minmax[1][j]) {
+          al.add(j);
+        }
+      }
+
+      // Check whether any eligible attributes have been found
+      if(al.size() == 0) {
+        return;
+      }
+      else {
+        // Randomly pick an attribute and split point
+        a = al.get(r.nextInt(al.size()));
+        q = (r.nextDouble() * (minmax[1][a] - minmax[0][a])) + minmax[0][a];
+
+        // Create sub trees
+        successors = new ArrayList<>(2);
+        for(int i = 0; i < 2; i++) {
+          ArrayModifiableDBIDs tmpData = DBIDUtil.newArray();
+          for(DBIDIter iditer = data.iter(); iditer.valid(); iditer.advance()) {
+            NumberVector vector = relation.get(iditer);
+            if((i == 0) && (vector.doubleValue(a) < q)) {
+              tmpData.add(iditer);
+            }
+            if((i == 1) && (vector.doubleValue(a) >= q)) {
+              tmpData.add(iditer);
+            }
+          }
+          successors.add(new Tree(tmpData, r, height + 1, maxHeight, relation));
+        }
+      }
+    }
+
+    /**
+     * Returns path length according to algorithm.
+     */
+    protected double pathLength(NumberVector vector) {
```

Comment:
An iterative version of this is likely more performant than a recursive version (which allocates stack frames).
Just use `while(successor != null) { depth++; successor = successors.get(...) }`